src/main/java/com/instaclustr/cassandra/auth/KerberosAuthenticator.java

```diff
@@ -0,0 +1,574 @@
+                        saslServer.evaluateResponse(response));
+            }
+            catch (PrivilegedActionException e)
+            {
+                logger.error("The SASL server could not evaluate the response sent by the client. " +
+                        "Check that the authentication mechanism is configured correctly, and that the client " +
+                        "is sending a valid SASL/{} response.", SASL_MECHANISM, e.getException());
+
+                // throw to client
+                throw new AuthenticationException("The SASL server could not evaluate the response sent by the client. " +
+                        "The server may not be configured correctly, or the response may be invalid.");
+            }
+        }
+
+        @Override
+        public boolean isComplete()
+        {
+            return saslServer.isComplete() && (this.authenticatedUser != null);
+        }
+
+        @Override
+        public AuthenticatedUser getAuthenticatedUser() throws AuthenticationException
+        {
+            if (!this.isComplete())
+                throw new AuthenticationException("SASL negotiation is not complete");
+
+            return this.authenticatedUser;
+        }
+    }
+
+    /**
+     * Check that a given pre-authenticated principal exists in Cassandra
+     *
+     * @param username Username of an externally-authenticated principal
+     * @return authenticated Cassandra user
+     */
+    private AuthenticatedUser getCassandraUser(String username) throws AuthenticationException
+    {
+        try
+        {
+            // This will throw an exception if the username does not exist
+            cache.get(username);
+
+            return new AuthenticatedUser(username);
+        }
+        catch (ExecutionException | UncheckedExecutionException e)
+        {
+            // the credentials were somehow invalid - either a non-existent role, or one without a defined password
+            if (e.getCause() instanceof NoSuchRoleException)
+                throw new AuthenticationException(String.format("Provided username %s is incorrect", username));
+
+            // an unanticipated exception occured whilst querying the credentials table
+            if (e.getCause() instanceof RequestExecutionException)
+            {
+                logger.trace("Error performing internal authentication", e);
+                throw new AuthenticationException(String.format("Error during authentication of user %s : %s", username, e.getMessage()));
+            }
+
+            throw new RuntimeException(e);
+        }
+    }
+
+    /**
+     * Query Cassandra role table for a given role
+     *
+     * @param roleName Cassandra role name
+     * @return role name
+     */
+    private String queryUserName(String roleName)
+    {
+        try
+        {
+            logger.debug("Querying role {}", roleName);
+            SelectStatement authenticationStatement = this.getRoleStatement;
```

Comment:
seems unnecessary. why not just call getRoleStatement below instead of reassigning to authenticationStatement?